src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -1882,7 +1904,9 @@ private ReadOnlyCollection<TKey> GetKeys()
             try
             {
                 AcquireAllLocks(ref locksAcquired);
-                List<TKey> keys = new List<TKey>();
+
+                int count = GetCountInternal();
+                List<TKey> keys = new List<TKey>(count);
```

Comment:
Is the count guaranteed not to change while all locks are held? If so, instead of a `List<TKey>` could `keys` be a `TKey[]` array to avoid the `List` allocation?